core/src/main/java/org/apache/druid/java/util/metrics/AllocationMetricCollector.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.java.util.metrics;
+
+import it.unimi.dsi.fastutil.longs.Long2LongMap;
+import it.unimi.dsi.fastutil.longs.Long2LongOpenHashMap;
+import org.apache.druid.java.util.common.logger.Logger;
+
+import java.lang.management.ThreadMXBean;
+import java.lang.reflect.Method;
+
+class AllocationMetricCollector
+{
+  private static final Logger log = new Logger(AllocationMetricCollector.class);
+
+  private static final int NO_DATA = -1;
+
+  private final Method getThreadAllocatedBytes;
+  private final ThreadMXBean threadMXBean;
+
+  private Long2LongMap previousResults;
+
+  AllocationMetricCollector(Method method, ThreadMXBean threadMXBean)
+  {
+    this.getThreadAllocatedBytes = method;
+    this.threadMXBean = threadMXBean;
+
+    previousResults = new Long2LongOpenHashMap();
+    previousResults.defaultReturnValue(NO_DATA);
+  }
+
+  /**
+   * Uses getThreadAllocatedBytes internally.
+   * Tests show the call to getThreadAllocatedBytes for a single thread ID out of 500 threads running takes around
+   * 9000 ns (in the worst case), which for 500 IDs should take 500*9000/1000/1000 = 4.5 ms to the max.
+   * AllocationMetricCollector takes linear time to calculate delta, for 500 threads it's negligible. One can call it as
+   * frequently as once a second, which is far too frequent for any known purpose to me.
```

Comment:
Instead of the passage "One can call it as frequently as once a second, which is far too frequent for any known purpose to me." please refer to the default emitting period and where it is configured.